File: wpilibc/src/test/native/cpp/util/interpolatable/TimeInterpolatableBufferTest.cpp
#include "frc/util/interpolatable/TimeInterpolatableBuffer.h"
#include "gtest/gtest.h"
#include <units/time.h>
#include "frc/geometry/Rotation2d.h"

TEST(TimeInterpolatableBufferTest, TestInterpolation) {
    frc::TimeInterpolatableBuffer<frc::Rotation2d> buffer{};

    buffer.addSample(0_s, frc::Rotation2d(0_rad));
    EXPECT_TRUE(buffer.getSample(0_s) == frc::Rotation2d(0_rad));
    buffer.addSample(1_s, frc::Rotation2d(1_rad));
    EXPECT_TRUE(buffer.getSample(0.5_s) == frc::Rotation2d(0.5_rad));
    EXPECT_TRUE(buffer.getSample(1_s) == frc::Rotation2d(1_rad));
    buffer.addSample(3_s, frc::Rotation2d(2_rad));
    EXPECT_TRUE(buffer.getSample(2_s) == frc::Rotation2d(1.5_rad));
}

File: wpimath/src/test/java/edu/wpi/first/math/interpolation/TimeInterpolatableBufferTest.java
package edu.wpi.first.math.interpolation;

import edu.wpi.first.wpilibj.geometry.Rotation2d;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class TimeInterpolatableBufferTest {
    @Test
    public void testBufferInterpolation() {
        TimeInterpolatableBuffer<Rotation2d> buffer = TimeInterpolatableBuffer.createBuffer(10);

        buffer.addSample(0, new Rotation2d());
        assertEquals(0, buffer.getSample(0).getRadians(), 0.001);
        buffer.addSample(1, new Rotation2d(1));
        assertEquals(0.5, buffer.getSample(0.5).getRadians(), 0.001);
        assertEquals(1.0, buffer.getSample(1.0).getRadians(), 0.001);
        buffer.addSample(3, new Rotation2d(2));
        assertEquals(1.5, buffer.getSample(2).getRadians(), 0.001);

        buffer.addSample(10.5, new Rotation2d(2));
        assertEquals(new Rotation2d(1), buffer.getSample(0));
    }
}
